keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+  ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,3,opt,name=objectSelector"`
+}
+```
+
+Example usage:
+
+```go
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ReplicaLimit struct {
+  MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,1,name=maxReplicas"`
+
+  Match Match `json:"match,omitempty" protobuf:"bytes,2,name=match"`
+}
+```
+
+For comparison (provided by @liggitt):
+
+- RBAC, priority & fairness, and admission webhook configurations all use
+ordered lists of rules, and first match allows/matches.
+- Audit configuration also uses an ordered list of rules, but can make a match
+exclude auditing... this approach could address the "ability to skip specific
+resource types" mentioned above.
+
+TODO: What exact matching mechanisms will be supported? Right now this KEP
+shows the use of the same match rules that admission webhooks use.
+
+TODO: Should policy definitions also have match criteria? Match criteria at the
+policy definition level could be used constrain what resources the policy may
+apply to across the entire cluster (like the above "kind constraint" but more
+generally). This would allow policy authors to constrain use of the policy to
+the resources it was designed for.
+
+TODO: Should "standalone policies" be supported? I.e. should it be possible to
+declare a policy definition that has no `spec.config` but has it's own match
+criteria? Right now a policy requires a minimum of three resources to do
+anything (policy definition, configuration CRD, at least one policy
+configuration).
+
+`MatchPolicy` will implicitly be `Equivalent`, and will not be configurable,
+but it will be possible to opt-out of the behavior by inspecting the
+GVK of the admitted resource in the CEL expression.
+
+xref:
+https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+
+#### Type safety
+
+To keep failure policy easy to reason about, and to continue to use CEL in a
+type-safe way we propose:
+
+- If a ValidatingAdmissionPolicy has a `spec.constraints` that matches a single
+  GVK, the CEL expression is allowed access to the full object in a typesafe
+  way. Otherwise, the CEL expression is allowed access to the metadata only.
+- If there are any type checking errors (or if the CRD for the matched GVK does
+  not exist):
+  - When a CRD is created/updated: The type check errors are detected by an
+    control loop watching the CRDs with an informer in the API server, and
+    reported in the status of the ValidatingAdmissionPolicy.
+  - When a policy is validated: runtime evaluations of the expression are
+    skipped, and the validation is reported as failed, with a type check error.
+
+The main disadvantage of this approach is that CEL rules that would have
```

Comment:
There is also the concern of version skew:

If k8s version N + 1 introduces a new Pod field, and a policy definition gets rewritten to accommodate that new field, then older versions of K8s cannot use that updated policy definition, even if it has other fixes, and even if the definition is written in such a way as to gracefully handle cases where the expected field is undefined.

Reply:
Dynamic code could guard access to fields with CEL `has(x.fieldname)` checks. This is a big ask for policy authors ("check every field you access to see what Kubernetes version it was introduced as and if that version is newer than the oldest version your policy supports, figure out how to write the policy so that it still works without that field..."). Granted, if the policy author does get this right, they get the benefit.  But for everyone else, they just loose type safety.

Reply:
To make this a bit more concrete, here is an example of a constraint template looking at seccomp policies that addresses the transition from using the annotation to using a field to define them:

https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/pod-security-policy/seccomp/template.yaml

Reply:
Thanks! I've added a use case and linked to this example.